host_applications/linux/apps/hello_pi/hello_mmal_encode/mmal_encode.c

```diff
@@ -0,0 +1,268 @@
+
+2. Redistributions in binary form must reproduce the above copyright notice,
+this list of conditions and the following disclaimer in the documentation
+and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its contributors
+may be used to endorse or promote products derived from this software without
+specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+// Image encoding example using MMAL
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+
+#include <bcm_host.h>
+#include <interface/mmal/mmal.h>
+#include <interface/mmal/mmal_format.h>
+#include <interface/mmal/util/mmal_default_components.h>
+#include <interface/mmal/util/mmal_component_wrapper.h>
+#include <interface/mmal/util/mmal_util_params.h>
+
+// Format for test image
+unsigned int WIDTH = 512;
+unsigned int HEIGHT = 512;
+MMAL_FOURCC_T INPUT_ENC = MMAL_ENCODING_RGBA;
+unsigned int BYTESPP = 4;
+uint32_t RED = 0xff;
+uint32_t GREEN = 0xff << 8;
+uint32_t BLUE = 0xff << 16;
+uint32_t ALPHA = 0xff << 24;
+
+VCOS_SEMAPHORE_T sem;
+
+// This callback and the above semaphore is used when waiting for
+// data to be returned.
+static void mmalCallback(MMAL_WRAPPER_T* mmal)
+{
+  vcos_semaphore_post(&sem);
+}
+
+// mmal_encode_test - Encode a test image and write to file
+void mmal_encode_test(MMAL_FOURCC_T encoding, // Encoding
+                      const char* filename) // File name
+{
+  MMAL_WRAPPER_T* mmal;
+  unsigned int inBufSize;
+  unsigned char* inBuf;
+  unsigned char* outBuf;
+  uint32_t* inPix; 
+  MMAL_PORT_T* portOut;
+  MMAL_PORT_T* portIn;
+  MMAL_BUFFER_HEADER_T* in;
+  MMAL_BUFFER_HEADER_T* out;
+  MMAL_STATUS_T status;
+  int eos = 0;
+  int sent = 0;
+  int outputWritten = 0;
+  int outFile;
+  int i, nw;
+  
+  fprintf(stdout, "Encoding test image %s\n", filename);
+
+  if (mmal_wrapper_create(&mmal, MMAL_COMPONENT_DEFAULT_IMAGE_ENCODER)
+      != MMAL_SUCCESS) {
+    fprintf(stderr, "Failed to create mmal component\n");
+    exit(1);
+  }
+  mmal->callback = mmalCallback;
+
+  // Configure input
+
+  portIn = mmal->input[0];
+  mmal->status = MMAL_SUCCESS;
+
+  if (portIn->is_enabled) {
```

Comment:
I'm not aware of any way for components to have enabled ports this soon after creation. Redundant? Or is this copy/paste from other examples? (In which case they ought to be amended).